### DISCUSSION TITLE
SciPy compatibility

### DISCUSSION THREAD

**Discussion by jonas-eschle** [zfit/zfit#670] [General]
Discussion about which features are needed.

**Comment by mdhaber** [comment]
Thanks for starting the discussion @jonas-eschle. 

I'm assuming the baseline is what you described in https://github.com/scipy/scipy/issues/23515#issuecomment-3275312918; i.e., that we would be able to work with SciPy distributions in zfit something like:

```python3
loss = zfit.loss.UnbinnedNLL(scipydist, data)
minimizer = zfit.minimize.ScipyBFGS()
result = minimizer.minimize(loss)
```

It sounds like you were already assuming `scipydist` would be a "new SciPy dist" - I agree it is fine to exclude `rv_generic` and focus on the new infrastructure.

One question that comes up is whether `scipydist` should be a class like `Normal` or an instance of one of these classes, say `X = Normal(mu=1, sigma=2)`. Some observations:
- The classes support multiple parameterizations; the parameterization is only selected when the class is instantiated. If `scipydist` were a class, we might need a way to select the parameterization that is to be used when fitting.
- An instance of a class will also have particular values of the parameters attached to it, which could serve as a guess.
- In the future, instances of these classes will be mutable, which will save a little overhead compared to instantiating the class every time you want to change the parameters.
- We might want to fit *transformed* instances (e.g. `stats.truncate(X, left, right)`) to data. (It is not as straightforward to transform the classes themselves.)

All these things suggest to me that `scipydist` should be an *instance* of one of these classes. Does that sound right?

---

With that out of the way, a few features SciPy currently has that we'd probably like to replicate are:
1. the ability to override the generic fitting procedure when there is a more efficient solution (e.g. analytical)
2. new loss functions (method of moments, maximum product spacing, and eventually method of L-moments)
3. fitting to [censored data](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.CensoredData.html#scipy.stats.CensoredData). See in [`fit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.fit.html#scipy.stats.rv_continuous.fit) how the log-likelihood function is modified to account for censored data.

I think that 1 would be particularly important to SciPy users in some cases. I know that `zfit` is intended to solve the more general problem, but is that a space you can see zfit getting into? How about the rest?

---

Finally, in https://github.com/scipy/scipy/issues/23515#issuecomment-3275312918, you wrote:
> You mentioned also the constraints; these are a bit tricky in a way: they can be implemented as penalties (which is easy to do in zfit), or for specific cases (as Matt mentioned). If possible, I found it more efficient to reformulate the problem, but this won't stop certain use-cases.

I wasn't quite sure I understood. There are a few kinds of constraints that arise frequently:
- simple bounds on distribution parameters (e.g. $a > 0$)
- simple inequality constraints between distribution parameters (e.g. $b > a$)
- more general but explicit constraints between distribution parameters (e.g. $b > a^2 + 1$)
- fully general constraints between distribution parameters ($a^2 + b^2 = 1$)
- constraints between distribution parameters and data (e.g. $a > x_i$)

The old SciPy infrastructure uses penalties to enforce some of these constraint, but ideally, it would be possible to treat them as `constraints` passed to the optimization algorithm. The SciPy objects have some knowledge of these constraints - could we work out a way for them to share this knowledge with `zfit` so `zfit` can use it?
